raft.go

```diff
@@ -1667,9 +1667,13 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 		hintIndex := min(m.Index, r.raftLog.lastIndex())
 		hintIndex = r.raftLog.findConflictByTerm(hintIndex, m.LogTerm)
 		hintTerm, err := r.raftLog.term(hintIndex)
-		if err != nil {
-			panic(fmt.Sprintf("term(%d) must be valid, but got %v", hintIndex, err))
+		if err != nil { // can be ErrCompacted or ErrUnavailable
+			hintTerm = 0
 		}
+		// NB: findConflictByTerm always returns an index <= its argument, so
+		// hintIndex <= min(m.Index, lastIndex()). This guarantees that the next
+		// append will be at a lower index. This is also the reason why it is safe
+		// to ignore a possible error above which, though, should never happen.
```

Comment:
I'm still uneasy about not populating the term hint. The reason the leader can handle this is because originally it was never populated, but we should consider this legacy behavior we'd want to migrate out of if we had a way to do so. If `findConflictByTerm` returned `(index, term, error)` and on success both `(index,term)` are populated, there would be fewer questions. And on error, we could log a warning and simply not respond, as is admissible in raft, and in particular for a rejection, or we could respond with the unadjusted `(index,term)`, so something like this:

```go
hintIndex := ...
hintTerm, err := ....
if err != nil {
  return // don't respond, log warning
}
adjHintIndex, adjHintTerm, err := r.raftLog.findConflictByTerm(...)
if err != nil { log warning but continue } else {
hintIndex, hintTerm = adjHintIndex, adjHintTerm
}
// respond
```

Reply:
As discussed in chat, it's more important to send a rejection `MsgAppResp` than get the hint exactly right.

There is an implicit expectation that a flow of `MsgApp` messages will eventually lead to a `MsgResp` (potentially with a reject). This is sort of a liveness bump for appends. If we introduce a condition under which a rejection `MsgResp` is not sent, we add some risk of ending up in a busy loop on the leader.